connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java

```diff
@@ -0,0 +1,547 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.connector.network.translators.world.block.entity;
+
+import com.github.steveice10.mc.protocol.data.game.world.block.value.PistonValue;
+import com.github.steveice10.mc.protocol.data.game.world.block.value.PistonValueType;
+import com.google.common.collect.ImmutableList;
+import com.nukkitx.math.vector.Vector3i;
+import com.nukkitx.nbt.NbtMap;
+import com.nukkitx.nbt.NbtMapBuilder;
+import com.nukkitx.protocol.bedrock.packet.BlockEntityDataPacket;
+import com.nukkitx.protocol.bedrock.packet.UpdateBlockPacket;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import org.geysermc.connector.network.session.GeyserSession;
+import org.geysermc.connector.network.translators.world.block.BlockStateValues;
+import org.geysermc.connector.network.translators.world.block.BlockTranslator;
+
+import java.util.*;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class PistonBlockEntity {
+    private final GeyserSession session;
+    private final Vector3i position;
+    private final PistonValue orientation;
+    private final boolean sticky;
+
+    private PistonValueType action;
+
+    /**
+     * A map of attached block positions to Java ids.
+     */
+    private final Object2IntMap<Vector3i> attachedBlocks = new Object2IntOpenHashMap<>();
+    /**
+     * A flattened array of the positions of attached blocks, stored in XYZ order.
+     */
+    private int[] flattenedAttachedBlocks = new int[0];
+
+    /**
+     * The position of the piston head
+     */
+    private float progress = 0.0f;
+    private float lastProgress = 0.0f;
+
+    private ScheduledFuture<?> updater;
+
+    private static final NbtMap AIR_TAG = BlockTranslator.BLOCKS.get(BlockTranslator.AIR).getCompound("block");
+    private static final List<Vector3i> ALL_DIRECTIONS = ImmutableList.of(Vector3i.from(1, 0, 0), Vector3i.from(0, 1, 0), Vector3i.from(0, 0, 1), Vector3i.from(-1, 0, 0), Vector3i.from(0, -1, 0), Vector3i.from(0, 0, -1));
```

Comment:
There's shortcuts to these in the form of `Vector3i.UP`, `Vector3i.DOWN`... perhaps those would be cleaner here?

Reply:
There's only shortcuts for `UP`, `RIGHT`, and `FORWARD`. I personally think mixing them with `Vector3i.from` may make it messier.